[DTensor] Register einsum sharding strategy for inference_mode support

stmcgovern opened the pull request:
Fixes #157631
Adds sharding strategy registration for `aten.einsum.default` to fix AssertionError when using einsum with DTensor in `inference_mode()`.
Adds tests



cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci @aditvenk @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @jerryzh168 @aditew01 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @Lucaskabela @chenyang78

stmcgovern committed:
[DTensor] Register einsum sharding strategy for inference_mode support

stmcgovern committed:
fix lint

stmcgovern commented:
resolved merge conflict.  Then switched target branch from viable/strict to main. Sorry for all the automatic review requests.

zpcore commented on torch/distributed/tensor/_ops/_matrix_ops.py:
```diff
@@ -267,6 +267,40 @@ def scaled_mm_strategy(op_schema: OpSchema) -> OpStrategy:
+    # einsum signature: einsum(equation, *operands)
+    # args_schema: [equation_str, OpStrategy1, OpStrategy2, ...]
+    equation = op_schema.args_schema[0]
+    assert isinstance(equation, str), (
```
Not sure if this is necessary. I guess if it is not a str, it should already error out somewhere.

zpcore commented on torch/distributed/tensor/_ops/_matrix_ops.py:
```diff
@@ -267,6 +267,40 @@ def scaled_mm_strategy(op_schema: OpSchema) -> OpStrategy:
+    variable-length operands as torch.einsum(equation, *operands).
+
+    Note: Currently supports einsum with up to 2 tensor operands and single output.
+    Multi-operand einsum (3+ tensors) is not yet supported and will raise an assertion.
```
Can you assert check if the op_schema[1] length is 2?

stmcgovern replied:
assert added

zpcore reviewed (commented):
Generally LGTM, left a few comments.

stmcgovern committed:
[DTensor] Address review feedback for einsum sharding strategy

  - Move test to test_tensor_ops.py per reviewer request
  - Use cast() for cleaner type narrowing (follows mm_strategy pattern)
  - Extract input strategies upfront for readability
  - Add redistribute cost computation
  - Fix test expectations: Partial() not Replicate() for contracting dims
  - Validate actual tensor operand count correctly

  Addresses feedback from @zpcore and @ezyang on PR #164995

ezyang commented:
Why do we not just decompose in this case? Saves you from having to write an entire new sharding strategy.

ezyang reviewed (approved):
ok

stmcgovern commented on test/distributed/tensor/test_dtensor_ops.py:
```diff
@@ -817,6 +817,78 @@ def run_opinfo_test(
     def assertEqualOnRank(self, x, y, msg=None, *, rank=0):
         self.assertEqual(x, y, msg)
 
+    def test_einsum_inference_mode(self):
```
done

stmcgovern committed:
use inference_mode multi-operand fallback

stmcgovern commented on torch/distributed/tensor/_ops/_matrix_ops.py:
```diff
@@ -267,6 +268,60 @@ def scaled_mm_strategy(op_schema: OpSchema) -> OpStrategy:
+    num_operands = len(tensor_operands.children)
+    assert num_operands <= 2, (
+        f"Currently only support einsum with up to 2 tensor operands, got {num_operands}. "
+        f"Multi-operand einsum (3+ tensors) is not yet supported."
```
Thanks @ezyang. Decomposition is great! I made a change so that it now falls back to decomposing when 3 or more operands are used. I don't want to change the logic of https://github.com/pytorch/pytorch/blob/2df2c316e2929519a5bad029936e9e37ed637b12/torch/distributed/tensor/_ops/_einsum_strategy.py#L31

I wonder if the decision to use decomposition or fused strategies is something that can be evaluated in the placement selection logic (available at trace time)? This might consider possible decompositions with computed_costs and compare to strategies and not just use decomposition as a fallback if no strategy is available.

ezyang reviewed (changes requested):
unfortunately needs a rebase now...

stmcgovern commented:
> unfortunately needs a rebase now...

No worries, I've rebased it. Were there other changes requested @ezyang ?

wconstab reviewed (commented):
Note that @weifengpy is working on cutting all the matmul ops over to new single dim einsum strat, it would be good to do the same for this.

wconstab reviewed (approved):
Seems ok to land though, and migrate later, if that's what you prefer.
